### PR TITLE
Push sdk

### PR DESCRIPTION
**Pull request by Boomatang**
# Push Notifications
## Prerequisite
* A valid `google-services.json` is required. This can be created in your firebase account.
* Under `main.assets` valid a `push-config.json` file is required. There is a sample in this folder.

## Running
1. Launch the app. In the logcat you should see the following message.
	`Main Activity: APP: Registration Succeeded`
2. With app in foreground. Send message from UPS. This will appear in the app a toast message. It may take a few moments.
3. With app in background or closed. Send message from UPS. This will be received as a notification in the top bar on the device. Once you click on the notification the app will open and the message will be toasted to the screen.

### CODE REVIEW

File: app/src/main/assets/sample-push-config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "pushServerURL": "pushServerURL (e.g http(s)//host:port/context)",
+  "android": {
+    "senderID": "senderID (e.g Google Project ID only for android)",
+    "variantID": "variantID (e.g. 1234456-234320)",
+    "variantSecret": "variantSecret (e.g. 1234456-234320)"
+  }
+}
```
Comment (wei-lee) [inline app/src/main/assets/sample-push-config.json:8]: what's the purpose of this file? Is it used as an example? If so we probably should update the README file to describe how users should add the config file for push.

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 2]
```diff
@@ -37,26 +47,55 @@
...
+    protected void onResume() {
+        super.onResume();
+        RegistrarManager.registerMainThreadHandler(this); // 1
+        RegistrarManager.unregisterBackgroundThreadHandler(NotifyingHandler.instance);
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        RegistrarManager.unregisterMainThreadHandler(this); // 2
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: what are the comments (`1`, `2`) for?

File: app/src/main/java/com/m/push/PushApplication.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                .asRegistrar();
+
+    }
+
+    // Accessor method for Activities to access the 'PushRegistrar' object
+    public PushRegistrar getPushRegistrar() {
+        return RegistrarManager.getRegistrar(PUSH_REGISTRAR_NAME);
+    }
+
+    public void addMessage(String message){
```
Comment (wei-lee) [inline app/src/main/java/com/m/push/PushApplication.java]: what's this function for? Is it just print out the message?

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java (hunk 2 above)
Comment (Boomatang) [reply app/src/main/java/com/m/androidNativeApp/MainActivity.java]: No idea. I might have put them in by mistake at some point and missed them when reading back over. I will remove them.

**Review by wei-lee — APPROVED** [review]
looks good overall, I just have a few questions.

File: app/src/main/assets/sample-push-config.json (hunk 1 above)
Comment (Boomatang) [reply app/src/main/assets/sample-push-config.json:8]: Yes it is to be used as an example. I will be put into the README, that's the next thing that we are working on.

File: app/src/main/java/com/m/push/PushApplication.java (hunk 3 above)
Comment (Boomatang) [reply app/src/main/java/com/m/push/PushApplication.java]: Yes it only prints out the message. It is been called when a notification is received by the device no matter if the app is in the foreground or background. We can remove it.

**Pull request merged**